### ISSUE TITLE
Label createWithTTF (Arabic font is broken) 

### ISSUE THREAD

**Issue by Kotaiba** [cocos2d/cocos2d-x#15321]
Hi, Arabic font is always broken on createWithTTF ,Arabic font only work on createWithSystemFont .
please look at the image how createWithTTF is broken and createWithSystemFont working fine 
![img_1121 copy](https://cloud.githubusercontent.com/assets/3635272/14062097/c544147a-f39b-11e5-87cb-6f79ac83076b.png)

**Comment by minggo** [comment]
Could provide some codes and the font file to reproduce it? And only this type of font can not work?

**Comment by Kotaiba** [comment]
I tried many fonts and all of them got the same issue .

```
//working fine 
auto lab=Label::createWithSystemFont("عربي","FrutigerLTArabic-65Bold.ttf" ,size);
this->addChild(lab,2);
```

```
//Bug
auto lab=Label::createWithTTF("عربي","FrutigerLTArabic-65Bold.ttf" ,size);
this->addChild(lab,2);
```

[FrutigerLTArabic-65Bold.ttf.zip](https://github.com/cocos2d/cocos2d-x/files/209815/FrutigerLTArabic-65Bold.ttf.zip)

**Comment by minggo** [comment]
Oh, what's the language. I think first you should make sure the string is utf-8 code, because labelTTF only support utf-8 string.

**Comment by Kotaiba** [comment]
How can i convert the string to utf-8 code ? Can you share code .

**Comment by minggo** [comment]
Because you write the string the source file, so you'd better make the source file be utf-8.

**Comment by Kotaiba** [comment]
Can you share a code how to do that ?

**Comment by minggo** [comment]
I mean you should save the .cpp file to utf-8 using an editor, such as sublime.

**Comment by Kotaiba** [comment]
ah, in my code i get the text from plist ,so i should save the plist to utf-8  .

**Comment by minggo** [comment]
But you wrote that:

```
//Bug
auto lab=Label::createWithTTF("عربي","FrutigerLTArabic-65Bold.ttf" ,size);
this->addChild(lab,2);
```

`("عربي"` is in a .cpp file.

**Comment by Kotaiba** [comment]
yes ,I've tried both and both same result ,I use both codes on my game .

**Comment by minggo** [comment]
FrutigerLTArabic-65Bold.ttf.zip it is a .ttf file doesn't include Arabic language, it just includes english word.

**Comment by Kotaiba** [comment]
Somehow the letters will not show but its arabic font . arabic letters will not show only english letter will show on arabic's fonts. but you can test its working fine .
<img width="275" alt="screen shot 2016-04-08 at 12 35 35 pm" src="https://cloud.githubusercontent.com/assets/3635272/14380097/7a94991e-fd86-11e5-9ddf-5d556f305f6e.png">

**Comment by minggo** [comment]
Sorry, i have not idea with it. So let @zilongshanren to take a look.

**Comment by larsven** [comment]
Try using cocos2d-x 3.11, our TTF label problems were fixed there

**Comment by Kotaiba** [comment]
@larsven the problem is not solved

**Comment by Kotaiba** [comment]
<img width="245" alt="screen shot 2016-07-17 at 4 28 25 pm" src="https://cloud.githubusercontent.com/assets/3635272/16900904/88948188-4c3b-11e6-8f5e-9d74b54ae40f.png">

**Comment by dumganhar** [comment]
@Kotaiba 
I tested it and found that the label created by `createWithTTF` will be shown as same as which in `Sublime Text Editor`.  Please refer to the following screenshot:
- `عربي` in Sublime Text:
  ![sublime](https://cloud.githubusercontent.com/assets/493372/19335731/032f2402-9138-11e6-87fb-d1eae95c81bd.png)
- `عربي` in Cocos2d-x:
  <img width="960" alt="font-test" src="https://cloud.githubusercontent.com/assets/493372/19335735/06fa7172-9138-11e6-8349-7a71b87c9b70.png">

**Comment by dumganhar** [comment]
@Kotaiba , I don't understand `Arabic font`, could you confirm that which one in cocos2d-x is correct? If System font is correct, why it's wrong in Sublime Text?

**Comment by Kotaiba** [comment]
@dumganhar the system font is correct one ,Sublime need to support Arabic in order for it to work . take a look at this post i think it will give you an idea . http://stackoverflow.com/questions/25603500/arabic-language-with-sublime-text-editor

**Comment by dumganhar** [comment]
@Kotaiba, I found this article (https://www.freetype.org/opentype/index.html).
Rendering Arabic font is more complex than other languages, we probably need to use `the FreeType OpenType Layout extension` for drawing `Arabic` letters.

**Comment by mantoun1984** [comment]
Any new about this ?

**Comment by ml0326660** [comment]
@minggo  @larsven  阿拉伯语文字添加ttf字体后显示与系统字体翻转的问题 现在有解决方案了没有？

**Comment by Kotaiba** [comment]
If anyone wanna quick fix , you can use http://omar84.com/docs/taw/arabic_writer.html . just get the Output and use it . only one issue you will face is the braking lines will be upside down in arabic . you can fix it by adding this inside 
`bool Label::multilineTextWrap(const std::function<int(const std::u16string&, int, int)>& nextTokenLen)`

```
----  nextTokenY -= (_lineHeight*_bmfontScale + lineSpacing);//change this to ->

   if (UserDefault::getInstance()->getBoolForKey("isArabic")){
                      nextTokenY += (_lineHeight*_bmfontScale + lineSpacing);
                       this->setPositionY(getPositionY()-(_lineHeight*_bmfontScale + lineSpacing));
                }else
                    nextTokenY -= (_lineHeight*_bmfontScale + lineSpacing);
```

and you need to set UserDefault::getInstance()->getBoolForKey("isArabic") true if arabic and false if not

**Comment by Arka161** [comment]
The issue still exists. Is there any workaround other than manually writing a parser, and then using the Arabic text?

**Comment by Arka161** [comment]
As mentioned, createWithTTF does make Arabic display in the reverse way, but createWithSystemFont does work. However, I do not need the system font and I would require a custom font.

**Comment by Arka161** [comment]
@Kotaiba The Arabic Writer does not work for all cases. For example, let's say we have an Arabic Text - "اللعب بدون اتصال بالانترنت". After processing it in the Arabic Writer URL that you have provided above, we get "ﺖﻧﺮﺘﻧﻻﺎﺑ ﻝﺎﺼﺗﺍ ﻥﻭﺪﺑ ﺐﻌﻠﻟﺍ". If we use that in Cocos2DX, we get a broken Arabic line of text.
I have tried reversing the line of text, however, it does not display the real thing on the screen. 

Would you have any suggestions on what to do?

**Comment by xulidong** [comment]
Is there a solution?

**Comment by Arvant** [comment]
i think if cocos2d-x add the harfbuzz lib with freetype the problem solved
https://harfbuzz.github.io/

**Comment by MohammadFakhreddin** [comment]
Hi, I wrote a fix for custom TTF . It is working for Persian but it still has some issues for some Arabic only characters. I appreciate If anybody with knowledge about Arabic can help me to fix remaining issues.
https://github.com/MohammadFakhreddin/cocos2dx-persian-arabic-support

**Comment by Arvant** [comment]
The best solution to this problem that respond to all languages is use of harfbuzz.

**Comment by MohammadFakhreddin** [comment]
Can you provide an example using this library?

**Comment by Arvant** [comment]
> Can you provide an example using this library?

https://github.com/anoek/ex-sdl-cairo-freetype-harfbuzz/

cocos2dx use freetype to get data from ttf file.and now need to reshape before rendering with harfbuzz

**Comment by MohammadFakhreddin** [comment]
Thank you I will check your suggestion to see if it is compatible to all cocos2d platforms

**Comment by Arvant** [comment]
> Thank you I will check your suggestion to see if it is compatible to all cocos2d platforms

yes ,harfbuzz used  in firefox,Qt,chrome in all platfroms
